File: javascript/network_benchmark.js
/**
 * Network Benchmark for Serverless Functions (Node.js)
 * Updated with new bandwidth test URL
 */

const http = require('http');
const https = require('https');
const dns = require('dns').promises;
const net = require('net');

async function benchmarkTcpLatency(host = '8.8.8.8', port = 53, iterations = 10) {
    const latencies = [];
    
    for (let i = 0; i < iterations; i++) {
        const start = Date.now();
        
        try {
            await new Promise((resolve, reject) => {
                const socket = net.connect(port, host, () => {
                    socket.end();
                    resolve();
                });
                socket.on('error', reject);
                socket.setTimeout(5000);
                socket.on('timeout', () => {
                    socket.destroy();
                    reject(new Error('Timeout'));
                });
            });
            
            const latency = Date.now() - start;
            latencies.push(latency);
        } catch (error) {
            console.error(`Connection failed: ${error.message}`);
            latencies.push(-1);
        }
    }
    
    const validLatencies = latencies.filter(l => l > 0);
    
    return {
        metric: 'tcp_latency',
        host,
        port,
        iterations,
        latencies_ms: validLatencies,
        min_ms: Math.min(...validLatencies),
        max_ms: Math.max(...validLatencies),
        avg_ms: validLatencies.reduce((a, b) => a + b, 0) / validLatencies.length,
        median_ms: validLatencies.sort((a, b) => a - b)[Math.floor(validLatencies.length / 2)],
        stdev_ms: calculateStdDev(validLatencies)
    };
}

async function benchmarkHttpLatency(url = 'http://httpbin.org/get', iterations = 10) {
    const latencies = [];
    
    for (let i = 0; i < iterations; i++) {
        const start = Date.now();
        
        try {
            await new Promise((resolve, reject) => {
                const client = url.startsWith('https') ? https : http;
                const req = client.get(url, (res) => {
                    let data = '';
                    res.on('data', chunk => data += chunk);
                    res.on('end', () => resolve(data));
                });
                req.on('error', reject);
                req.setTimeout(10000);
                req.on('timeout', () => {
                    req.destroy();
                    reject(new Error('Timeout'));
                });
            });
            
            const latency = Date.now() - start;
            latencies.push(latency);
        } catch (error) {
            console.error(`HTTP request failed: ${error.message}`);
            latencies.push(-1);
        }
    }
    
    const validLatencies = latencies.filter(l => l > 0);
    
    return {
        metric: 'http_latency',
        url,
        iterations,
        latencies_ms: validLatencies,
        min_ms: Math.min(...validLatencies),
        max_ms: Math.max(...validLatencies),
        avg_ms: validLatencies.reduce((a, b) => a + b, 0) / validLatencies.length,
        median_ms: validLatencies.sort((a, b) => a - b)[Math.floor(validLatencies.length / 2)],
        stdev_ms: calculateStdDev(validLatencies)
    };
}

async function benchmarkDnsResolution(domain = 'google.com', iterations = 10) {
    const latencies = [];
    
    for (let i = 0; i < iterations; i++) {
        const start = Date.now();
        
        try {
            await dns.resolve4(domain);
            const latency = Date.now() - start;
            latencies.push(latency);
        } catch (error) {
            console.error(`DNS resolution failed: ${error.message}`);
            latencies.push(-1);
        }
    }
    
    const validLatencies = latencies.filter(l => l > 0);
    
    return {
        metric: 'dns_resolution',
        domain,
        iterations,
        latencies_ms: validLatencies,
        min_ms: Math.min(...validLatencies),
        max_ms: Math.max(...validLatencies),
        avg_ms: validLatencies.reduce((a, b) => a + b, 0) / validLatencies.length,
        median_ms: validLatencies.sort((a, b) => a - b)[Math.floor(validLatencies.length / 2)]
    };
}

async function benchmarkBandwidth(url = 'https://hagimont.freeboxos.fr/hagimont/software/resteasy-jaxrs-3.0.9.Final-all.zip') {
    const downloadTimes = [];
    let fileSizeMb = null;
    
    for (let i = 0; i < 3; i++) {
        const start = Date.now();
        
        try {
            const data = await new Promise((resolve, reject) => {
                const client = url.startsWith('https') ? https : http;
                const req = client.get(url, (res) => {
                    const chunks = [];
                    res.on('data', chunk => chunks.push(chunk));
                    res.on('end', () => resolve(Buffer.concat(chunks)));
                });
                req.on('error', reject);
                req.setTimeout(30000);
            });
            
            const downloadTime = (Date.now() - start) / 1000; // Convert to seconds
            downloadTimes.push(downloadTime);
            
            // Calculate file size
            const sizeBytes = data.length;
            fileSizeMb = sizeBytes / (1024 * 1024);
            
        } catch (error) {
            console.error(`Download failed: ${error.message}`);
            return { 
                metric: 'bandwidth',
                url,
                error: error.message 
            };
        }
    }
    
    const avgTime = downloadTimes.reduce((a, b) => a + b, 0) / downloadTimes.length;
    const avgBandwidth = (fileSizeMb * 8) / avgTime; // Mbps
    
    return {
        metric: 'bandwidth',
        url,
        file_size_mb: Math.round(fileSizeMb * 100) / 100,
        download_times_s: downloadTimes,
        avg_download_time_s: avgTime,
        bandwidth_mbps: avgBandwidth,
        min_time_s: Math.min(...downloadTimes),
        max_time_s: Math.max(...downloadTimes)
    };
}

function calculateStdDev(values) {
    const avg = values.reduce((a, b) => a + b, 0) / values.length;
    const squareDiffs = values.map(value => Math.pow(value - avg, 2));
    const avgSquareDiff = squareDiffs.reduce((a, b) => a + b, 0) / squareDiffs.length;
    return Math.sqrt(avgSquareDiff);
}

async function runAllBenchmarks() {
    const results = {
        timestamp: Date.now(),
        benchmarks: {}
    };
    
    console.log('Running TCP Latency Benchmark...');
    results.benchmarks.tcp_latency = await benchmarkTcpLatency();
    
    console.log('Running HTTP Latency Benchmark...');
    results.benchmarks.http_latency = await benchmarkHttpLatency();
    
    console.log('Running DNS Resolution Benchmark...');
    results.benchmarks.dns_resolution = await benchmarkDnsResolution();
    
    console.log('Running Bandwidth Benchmark...');
    results.benchmarks.bandwidth = await benchmarkBandwidth();
    
    return results;
}

async function main(params) {
    return await runAllBenchmarks();
}

if (require.main === module) {
    runAllBenchmarks().then(results => {
        console.log(JSON.stringify(results, null, 2));
    }).catch(error => {
        console.error('Error:', error);
        process.exit(1);
    });
}

exports.main = main;

File: python/network_benchmark.py
#!/usr/bin/env python3
"""
Network Benchmark for Serverless Functions
Updated with new bandwidth test URL
"""
import time
import socket
import json
import statistics
from urllib.request import urlopen
from urllib.error import URLError

def benchmark_tcp_latency(host="8.8.8.8", port=53, iterations=10):
    """TCP Connection Latency Benchmark"""
    latencies = []
    
    for i in range(iterations):
        start = time.time()
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(5)
            sock.connect((host, port))
            sock.close()
            latency = (time.time() - start) * 1000
            latencies.append(latency)
        except Exception as e:
            print(f"Connection failed: {e}")
            latencies.append(-1)
    
    valid_latencies = [l for l in latencies if l > 0]
    
    return {
        "metric": "tcp_latency",
        "host": host,
        "port": port,
        "iterations": iterations,
        "latencies_ms": valid_latencies,
        "min_ms": min(valid_latencies) if valid_latencies else 0,
        "max_ms": max(valid_latencies) if valid_latencies else 0,
        "avg_ms": statistics.mean(valid_latencies) if valid_latencies else 0,
        "median_ms": statistics.median(valid_latencies) if valid_latencies else 0,
        "stdev_ms": statistics.stdev(valid_latencies) if len(valid_latencies) > 1 else 0
    }

def benchmark_http_latency(url="http://httpbin.org/get", iterations=10):
    """HTTP Request Latency Benchmark"""
    latencies = []
    
    for i in range(iterations):
        start = time.time()
        try:
            response = urlopen(url, timeout=10)
            response.read()
            latency = (time.time() - start) * 1000
            latencies.append(latency)
        except URLError as e:
            print(f"HTTP request failed: {e}")
            latencies.append(-1)
    
    valid_latencies = [l for l in latencies if l > 0]
    
    return {
        "metric": "http_latency",
        "url": url,
        "iterations": iterations,
        "latencies_ms": valid_latencies,
        "min_ms": min(valid_latencies) if valid_latencies else 0,
        "max_ms": max(valid_latencies) if valid_latencies else 0,
        "avg_ms": statistics.mean(valid_latencies) if valid_latencies else 0,
        "median_ms": statistics.median(valid_latencies) if valid_latencies else 0,
        "stdev_ms": statistics.stdev(valid_latencies) if len(valid_latencies) > 1 else 0
    }

def benchmark_bandwidth(url="https://hagimont.freeboxos.fr/hagimont/software/resteasy-jaxrs-3.0.9.Final-all.zip"):
    """
    Bandwidth Benchmark
    Downloads file and measures throughput
    """
    download_times = []
    file_size_mb = None
    
    for i in range(3):
        start = time.time()
        try:
            response = urlopen(url, timeout=30)
            data = response.read()
            download_time = time.time() - start
            download_times.append(download_time)
            
            # Calculate actual file size
            size_bytes = len(data)
            file_size_mb = size_bytes / (1024 * 1024)
            
        except URLError as e:
            print(f"Download failed: {e}")
            return {
                "metric": "bandwidth",
                "url": url,
                "error": str(e)
            }
    
    avg_time = statistics.mean(download_times)
    avg_bandwidth = (file_size_mb * 8) / avg_time  # Mbps
    
    return {
        "metric": "bandwidth",
        "url": url,
        "file_size_mb": round(file_size_mb, 2) if file_size_mb else 0,
        "download_times_s": download_times,
        "avg_download_time_s": avg_time,
        "bandwidth_mbps": avg_bandwidth,
        "min_time_s": min(download_times),
        "max_time_s": max(download_times)
    }

def benchmark_dns_resolution(domain="google.com", iterations=10):
    """DNS Resolution Latency"""
    latencies = []
    
    for i in range(iterations):
        start = time.time()
        try:
            socket.gethostbyname(domain)
            latency = (time.time() - start) * 1000
            latencies.append(latency)
        except socket.gaierror as e:
            print(f"DNS resolution failed: {e}")
            latencies.append(-1)
    
    valid_latencies = [l for l in latencies if l > 0]
    
    return {
        "metric": "dns_resolution",
        "domain": domain,
        "iterations": iterations,
        "latencies_ms": valid_latencies,
        "min_ms": min(valid_latencies) if valid_latencies else 0,
        "max_ms": max(valid_latencies) if valid_latencies else 0,
        "avg_ms": statistics.mean(valid_latencies) if valid_latencies else 0,
        "median_ms": statistics.median(valid_latencies) if valid_latencies else 0
    }

def run_all_benchmarks():
    """Run all network benchmarks"""
    results = {
        "timestamp": time.time(),
        "benchmarks": {}
    }
    
    print("Running TCP Latency Benchmark...")
    results["benchmarks"]["tcp_latency"] = benchmark_tcp_latency()
    
    print("Running HTTP Latency Benchmark...")
    results["benchmarks"]["http_latency"] = benchmark_http_latency()
    
    print("Running DNS Resolution Benchmark...")
    results["benchmarks"]["dns_resolution"] = benchmark_dns_resolution()
    
    print("Running Bandwidth Benchmark...")
    results["benchmarks"]["bandwidth"] = benchmark_bandwidth()
    
    return results

def main(args):
    """OpenWhisk entry point"""
    return run_all_benchmarks()

if __name__ == "__main__":
    results = run_all_benchmarks()
    print(json.dumps(results, indent=2))